instrumentation/elasticsearch/elasticsearch-rest-common/javaagent/src/main/java/io/opentelemetry/javaagent/instrumentation/elasticsearch/rest/ElasticsearchClientAttributeExtractor.java

```diff
@@ -0,0 +1,124 @@
+
+import io.opentelemetry.api.common.AttributeKey;
+import io.opentelemetry.api.common.AttributesBuilder;
+import io.opentelemetry.context.Context;
+import io.opentelemetry.instrumentation.api.instrumenter.AttributesExtractor;
+import io.opentelemetry.semconv.trace.attributes.SemanticAttributes;
+import java.net.Inet6Address;
+import java.util.ArrayList;
+import java.util.List;
+import java.util.Map;
+import java.util.concurrent.ConcurrentHashMap;
+import java.util.regex.Matcher;
+import java.util.regex.Pattern;
+import javax.annotation.Nullable;
+import org.apache.http.HttpHost;
+import org.elasticsearch.client.Response;
+
+public class ElasticsearchClientAttributeExtractor
+    implements AttributesExtractor<ElasticsearchRestRequest, Response> {
+  private static final Pattern pathPartNamesPattern = Pattern.compile("\\{([^}]+)}");
+  private final Map<String, Pattern> regexPatternMap = new ConcurrentHashMap<>();
+  private final Map<String, List<String>> routePathPartNames = new ConcurrentHashMap<>();
+
+  @Override
+  public void onStart(
+      AttributesBuilder attributes, Context parentContext, ElasticsearchRestRequest request) {
+    internalSet(attributes, SemanticAttributes.HTTP_METHOD, request.getMethod());
+    setPathPartsAttributes(attributes, request);
+  }
+
+  @Override
+  public void onEnd(
+      AttributesBuilder attributes,
+      Context context,
+      ElasticsearchRestRequest request,
+      @Nullable Response response,
+      @Nullable Throwable error) {
+    if (response != null) {
+      HttpHost host = response.getHost();
+      String uri = response.getRequestLine().getUri();
+      uri = uri.startsWith("/") ? uri : "/" + uri;
+      String url = host.toURI() + uri;
+
+      internalSet(attributes, SemanticAttributes.HTTP_URL, url);
+
+      if (host.getAddress() != null) {
+        internalSet(
+            attributes, SemanticAttributes.NET_PEER_NAME, host.getAddress().getHostAddress());
+        internalSet(attributes, SemanticAttributes.NET_PEER_PORT, (long) host.getPort());
+        if (host.getAddress() instanceof Inet6Address) {
+          internalSet(attributes, SemanticAttributes.NET_SOCK_FAMILY, "inet6");
+        }
+      }
+    }
+  }
+
+  private void setPathPartsAttributes(
+      AttributesBuilder attributes, ElasticsearchRestRequest request) {
+    ElasticsearchEndpointDefinition endpointDef = request.getEndpointDefinition();
+    if (endpointDef == null) {
+      return;
+    }
+    String[] availableRoutes = endpointDef.getRoutes();
+    if (availableRoutes == null || availableRoutes.length == 0) {
+      return;
+    }
+    String urlPath = request.getEndpoint();
+    boolean routeFound = false;
+    for (int i = 0; i < availableRoutes.length && !routeFound; i++) {
+      String route = availableRoutes[i];
+      if (route.contains("{")) {
+        Matcher matcher = matchUrl(route, urlPath);
+        if (matcher.find()) {
+          List<String> pathPartNames = routePathPartNames.get(route);
```

Comment:
lets assume that some other thread is also calling `matchUrl` then this thread could observer `regexPatternMap` being update, but `pathPartNamesPattern` not yet updated

Reply:
Refactored that logic and also added comments that explain what happens here.